### ISSUE TITLE
Mypy confuses enum constants with enum values when used as a dictionary key

### ISSUE THREAD

**Issue by thomascellerier** [python/mypy#6712]
Mypy seems to be confusing enum constants with enum values when used as a disctionary key.
This results in both false positives and false negatives.
I made a minimal example of a false negative here but this is stopping me from type checking a bigger program by reporting false positives.

* Are you reporting a bug, or opening a feature request?

This is a bug

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```Python
from typing import Dict
from enum import Enum

class Things(Enum):
    Foo = 1
    Bar = 2

things: Dict[Things, str] = {
    Things.Foo: 'foo'
}

print(things[Things.Foo])
print(things[Things.Foo.value])
```

* What is the actual behavior/output?

The program typechecks but fails at runtime.

```
$ mypy example.py
$ python3 example.py 
foo
Traceback (most recent call last):
  File "example.py", line 14, in <module>
    print(things[Things.Foo.value])
KeyError: 1
```

* What is the behavior/output you expect?

The program should fail typechecking.

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

Python 3.6.7 (Ubuntu 18.04).
Mypy 0.701 (pip).
Same using mypy-0.710+dev.1b9b2b36a03d364f1bd4fdd6233c9f794dd3f985.

* What are the mypy flags you are using? (For example --strict-optional)

None.

**Comment by Michael0x2a** [comment]
The root issue is because mypy currently does not know what the type of `SomeEnum.FOO.value` ought to be and defaults to assuming that value is of type `Any`:

```
from enum import Enum

class Test(Enum):
    FOO = 1

reveal_type(Test.FOO)        # Revealed type is 'Test'
reveal_type(Test.FOO.value)  # Revealed type is 'Any', instead of 'int'.
```

I'm probably going to implement a plugin to handle this case as part of the ongoing enums and literal types project.

**Comment by ltworf** [comment]
@thomascellerier perhaps you could provide an example of code that fails to validate but should, as well?

**Comment by thomascellerier** [comment]
I can't reproduce a minimal example but I have actual problems with this where it assumes the type is `str` or `int` and not `Any`:
```Python
testy/command.py:99: error: Invalid index type "int" for "Dict[OS, Callable[[Popen], None]]"; expected type "OS"
testy/command.py:112: error: Invalid index type "int" for "Dict[OS, Callable[[Popen], None]]"; expected type "OS"
testy/command.py:163: error: Incompatible default for argument "stdout_dest" (default has type "int", argument has type "Destination")
testy/command.py:163: error: Incompatible default for argument "stderr_dest" (default has type "int", argument has type "Destination")
```

A relevant code snippet:
```Python
class OS(Enum):
    Linux = 1
    Darwin = 2
    Windows = 3
    Android = 4
    iOS = 5

_kill_mapping: Dict[platform.OS, Callable[[subprocess.Popen], None]] = {
    platform.OS.Darwin: kill_unix,
    platform.OS.Linux: kill_unix,
    platform.OS.Windows: kill_windows,
    platform.OS.Android: kill_mobile,
    platform.OS.iOS: kill_mobile
}
try:
    kill = _kill_mapping[platform.os]
except KeyError:
```

And another:
```Python
class Destination(Enum):
    REPORTING = 1
    BUFFER = 2

    def spawn_elevated(self, cmd: list, name: Optional[str] = None, stdin_src: Optional[Union[str, bytes]] = None,
                       stdout_dest: Destination = Destination.REPORTING,
                       stderr_dest: Destination = Destination.REPORTING,
                       *, env: Optional[Dict[str, str]] = None, user: Optional[str] = None) -> 'Process':
```

**Comment by gvanrossum** [comment]
You're welcome to continue the discussion but please post self-contained code snippets. I don't know what to do with your examples (which on their own don't work at all) and matching the errors you quote with the example code.

**Comment by thomascellerier** [comment]
Sorry about that I couldn't find a minimal example.
After looking into it more the root my other issues seems completely unrelated to enums.
The initial issue reported still stands and @Michael0x2a 's response explains the situation.

**Comment by gvanrossum** [comment]
Thanks, then I'm closing this issue.

**Comment by ilevkivskyi** [comment]
@Michael0x2a 
> I'm probably going to implement a plugin to handle this case as part of the ongoing enums and literal types project.

Great idea! It will be another nice test for usability of our plugin system.

**Comment by jayceslesar** [comment]
Here is a minimal example:

```py
from enum import Enum


class Example(Enum):
        A = 1
        B = 2
        C = 3

        def example_function(self) -> str:
            d = {
                self.A: 'a',
                self.B: 'b',
                self.C: 'c'
            }
            return d[self]
```
Running `mypy example_bad.py` produces `example_bad.py:15: error: Invalid index type "Example" for "Dict[int, str]"; expected type "int"`
